### PR TITLE
[Feature] Support TTA

### PR DESCRIPTION
**Pull request by hhaAndroid**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

based on mmdet dev 3.x.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by RangeKing — COMMENTED** [review]
79fbccaca6b562105856333f705f1bf
better to add some TTA performance in README?

File: configs/_base_/det_p5_tta.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# TODO: Need to solve the problem of multiple file_client_args parameters
+# _file_client_args = dict(
```
Comment (PeterH0323) [inline configs/_base_/det_p5_tta.py:2]: 这个可以删掉

File: configs/_base_/det_p5_tta.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#         './data/': 's3://openmmlab/datasets/detection/',
+#         'data/': 's3://openmmlab/datasets/detection/'
+#     }))
+_file_client_args = dict(backend='disk')
+
+tta_model = dict(
+    type='mmdet.DetTTAModel',
+    tta_cfg=dict(nms=dict(type='nms', iou_threshold=0.65), max_per_img=300))
+
+img_scales = [(640, 640), (320, 320), (960, 960)]
```
Comment (PeterH0323) [inline configs/_base_/det_p5_tta.py:14]: config 要不要像优化过的 config 那样分段加下注释？

File: configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py [hunk 3]
```diff
@@ -232,3 +234,36 @@
 test_evaluator = val_evaluator
 
 train_cfg = dict(max_epochs=max_epochs)
+
```
Comment (PeterH0323) [inline configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py:237]: 这个 config 的 tta 配置能否继承？

File: configs/yolov5/yolov5_s-p6-v62_syncbn_fast_8xb16-300e_coco.py [hunk 4]
```diff
@@ -26,6 +26,7 @@
 obj_level_weights = [4.0, 1.0, 0.25, 0.06]
 affine_scale = 0.5  # YOLOv5RandomAffine scaling ratio
 
+tta_img_scales = [(1280, 1280), (1024, 1024), (1536, 1536)]
```
Comment (PeterH0323) [inline configs/yolov5/yolov5_s-p6-v62_syncbn_fast_8xb16-300e_coco.py:29]:
```suggestion
# Config for Test Time Augmentation. (TTA)
# You only need to specify `--tta` when testing to enable.
tta_img_scales = [(1280, 1280), (1024, 1024), (1536, 1536)]
```

File: configs/yolov7/yolov7_w-p6_syncbn_fast_8x16b-300e_coco.py [hunk 5]
```diff
@@ -138,6 +145,37 @@
     dataset=dict(pipeline=test_pipeline, batch_shapes_cfg=batch_shapes_cfg))
 test_dataloader = val_dataloader
 
-# The only difference between P6 and P5 in terms of
-# hyperparameters is lr_factor
-default_hooks = dict(param_scheduler=dict(lr_factor=0.2))
+default_hooks = dict(param_scheduler=dict(lr_factor=lr_factor))
```
Comment (PeterH0323) [inline configs/yolov7/yolov7_w-p6_syncbn_fast_8x16b-300e_coco.py:148]: tta config 能否继承？

File: docs/zh_cn/common_usage/tta.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
+# 测试时增强相关说明
+
+## 测试时增强 TTA
+
+MMYOLO 在 v0.5.0+ 版本中增加对 TTA 的支持，用户可以在进行评估时候指定 `--tta` 参数使能。 以 `YOLOv5-s` 为例，其单卡 TTA 测试命令为：
+
+```shell
+python tools/test.py configs/yolov5/yolov5_n-v61_syncbn_fast_8xb16-300e_coco.py https://download.openmmlab.com/mmyolo/v0/yolov5/yolov5_n-v61_syncbn_fast_8xb16-300e_coco/yolov5_n-v61_syncbn_fast_8xb16-300e_coco_20220919_090739-b804c1ad.pth  --tta
```
Comment (PeterH0323) [inline docs/zh_cn/common_usage/tta.md:8]: 我建议每个入参都换下行，看起来容易看一些

```bash
python demo.py arg1 \
               arg2 \
               ...
```

File: configs/_base_/det_p5_tta.py (hunk 1 above)
Comment (hhaAndroid) [reply configs/_base_/det_p5_tta.py:2]: 要保留，有时候有用

File: configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py (hunk 3 above)
Comment (hhaAndroid) [reply configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py:237]: 可以继承呀

Comment (hhaAndroid) [reply configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py:237]: tta_pipeline 是一个list，即使继承其实也是全部重写

**Pull request merged**